[wip] merge on green bot v2

zengk95 opened the pull request:
Second version of Green Bot that spurned from the conversations in https://github.com/pytorch/test-infra/pull/302

zengk95 committed:
Revert "[nnc] Strides to Tensor (#72962)"

This reverts commit 939060925f28c9498da42225f216d838e1f7f4ca.

H-Huang committed:
[RPC small change] Improve logging from 'unknown destination worker'

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75811

Approved by: https://github.com/mrshenli

Commit:
[Testing CI] test exact layout on nvfuser tests

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76393

Approved by: https://github.com/jjsjann123, https://github.com/davidberard98

zengk95 committed:
Merge branch 'master' of https://github.com/pytorch/pytorch

jbschlosser committed:
Support for tensor subclasses as parameters

Pull Request resolved: https://github.com/pytorch/pytorch/pull/73459

Approved by: https://github.com/ezyang, https://github.com/albanD

zengk95 committed:
Merge branch 'master' of https://github.com/pytorch/pytorch

zengk95 committed:
updated try merge

zengk95 committed:
updated try merge

zengk95 committed:
udpated error

zengk95 committed:
udpated error

henrytwo committed:
Enable LTC Input/Output Mapping (#75828)

Summary:
This PR enables Input/Output aliasing for Lazy Tensor Core. `SetUpAlias` is a virtual function that can be overridden in a vendor's custom `LoweringContext` implementation.

The return type of `LoweringContext::GetResultShape` has also been updated to return a `c10::optional` value, since `GetResultShape` isn't currently implemented for the TorchScript backend.

The changes here mirror the interface used by `torch_xla`: https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1548-L1549

cc: antoniojkim ke1337 wconstab silvasean

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75828

Reviewed By: Krovatkin

Differential Revision: D35952593

Pulled By: wconstab

fbshipit-source-id: e20b11e44e0e1beda1b1c47aa3a8b611afd97b7f
(cherry picked from commit bcbc9ef01ef8eb84667e5c42edc10d38d5d78395)

janeyx99 commented on .github/scripts/trymerge.py:
```diff
@@ -652,6 +653,25 @@ def merge_into(self, repo: GitRepo, *, force: bool = False, dry_run: bool = Fals
+                raise RuntimeError(msg)
+
+            try:
+                self.merge_into(repo, dry_run=dry_run)
```
Wouldn't this still post confusing comments on the PR that merge failed every 60s? 

You might need to change some logic in the other parts of the script if ON_GREEN is true

zengk95 replied:
I don't believe so. The merge_into method doesn't contain anything that would post a comment. It only posts a comment when they catch the exception thrown from within it like on line L856.

janeyx99 commented on .github/workflows/trymerge.yml:
```diff
@@ -29,9 +29,12 @@ jobs:
           if [ -n "${FORCE}" ]; then
             python3 .github/scripts/trymerge.py --force "${PR_NUM}"
+          elif [-n "${ON_GREEN}"]; then
+            python3. github/scripts/trymerge.py --onGreen "${PR_NUM}"
```
I believe you would need a corresponding flag in parse_args(), and the convention is usually --on-green (vs the camel case)

pruthvistony committed:
[ROCm] Skipping few multiprocess test

- Found it is failing on ROCm 5.1.1, will be enabled back
  as soon as it is fixed.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76402
Approved by: https://github.com/jeffdaily, https://github.com/seemethere

zengk95 committed:
updated files

zengk95 committed:
updated files

Commit:
Revert D35860266: [pytorch][PR] Update torch::lazy::BackendDevice to have a new default ordinal

Test Plan: revert-hammer

Differential Revision:
D35860266 (https://github.com/pytorch/pytorch/commit/f9d07ae6449224bdcb6eb69044a33f0fb5780adf)

Original commit changeset: 554ebe16a068

Original Phabricator Diff: D35860266 (https://github.com/pytorch/pytorch/commit/f9d07ae6449224bdcb6eb69044a33f0fb5780adf)

fbshipit-source-id: 325c54aa2e87e51134115213352b3d33a81b7edf
(cherry picked from commit bbd74bf34a534d1b87aadff9790038e3dbbfa9c8)

zengk95 committed:
fix period

zengk95 committed:
fix period

zengk95 committed:
fixed lint

zengk95 committed:
fixed lint

cbalioglu committed:
Add new dispatch keys for Fake Tensor and Deferred Module Initialization

Thanks to @bdhirsh's work, we now have room for new dispatch keys in `DispatchKey` enum. This PR adds two new keys for out-of-core [Fake Tensor](https://pytorch.org/torchdistx/latest/fake_tensor.html) and [Deferred Module Initialization](https://pytorch.org/torchdistx/latest/deferred_init.html) features.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76139
Approved by: https://github.com/bdhirsh

kulinseth committed:
Add "mps" device to PyTorch framework.

Remove the "mlc" device for Mac platforms.

This commit will be followed up with:

* adding MPS runtime components
* PyTorch ops for MPS device

Fixes #ISSUE_NUMBER

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76291
Approved by: https://github.com/albanD

Commit:
[PyTorch] Use native serial stack when there is only 1 thread (#76399)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76399

Does not make sense to hit the non-serial kernel if intra op parallelism is off.
ghstack-source-id: 154934866

Test Plan: Existing unit tests

Reviewed By: tenpercent

Differential Revision: D35946012

fbshipit-source-id: 9ae12e267826b4da96b0a105c34260bc4ac91135
(cherry picked from commit f61eb10a754cab5491ec972b72aa5e92e0638e50)

cpuhrsch committed:
Port to_padded_tensor CUDA kernel from pytorch/nestedtensor

This PR adds a custom CUDA kernel to pad NestedTensors between dimension 2 and 4.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76157
Approved by: https://github.com/ngimel

jeffdaily committed:
[ROCm][GHA] keep docker images for at most 1 day

Fixes #76413.  ROCm runners take longer to pull docker images than other runners.  Try to cache images for at most 1 day to improve workflow throughput.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76415
Approved by: https://github.com/janeyx99, https://github.com/clee2000, https://github.com/seemethere

jeffdaily committed:
[ROCm] persist test results even on failure

Fixes #75169.  Default shell for workflow steps will fail fast.  Split ROCm test execution and test result copy into two steps so that test results are always persisted.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76405
Approved by: https://github.com/seemethere

ninginthecloud committed:
[DataPipe] Update `mux` data pipe (#76384)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76384

OSS issue discussion: https://github.com/pytorch/data/issues/346
This diff updates `mux` and `mux_longest` data pipe.
`mux`: Yields one element at a time from each of the input Iterable DataPipes (functional name: ``mux``). As in, one element from the 1st input DataPipe, then one element from the 2nd DataPipe in the next iteration, and so on. It ends when the shortest input DataPipe is exhausted.

`mux` example:

```
>>> from torchdata.datapipes.iter import IterableWrapper
>>> dp1, dp2, dp3 = IterableWrapper(range(3)), IterableWrapper(range(10, 15)), IterableWrapper(range(20, 25))
>>> list(dp1.mux(dp2, dp3))
[0, 10, 20, 1, 11, 21, 2, 12, 22]
```

Test Plan:
buck test mode/dev //pytorch/data/test:tests -- --exact 'pytorch/data/test:tests - test_mux_longest_iterdatapipe (test_datapipe.TestDataPipe)'

https://www.internalfb.com/intern/testinfra/testrun/3096224791148107

Reviewed By: ejguan

Differential Revision: D35799965

fbshipit-source-id: 320e71a342ec27e6e9200624aad42f4b99f97c3a
(cherry picked from commit 741ed595275df6c05026ed6f0e78d7052328fb7d)

mikaylagawarecki committed:
Prototype _index_reduce  (CPU-only)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75981

Approved by: https://github.com/cpuhrsch

anjali411 committed:
Add __all__ for torch.cuda.memory

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76490

Approved by: https://github.com/albanD

davidberard98 committed:
[NVFuser] Reduce accuracy threshold for fp16 nvfuser opinfos

Re-enable & reduce accuracy threshold

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76495

Approved by: https://github.com/mruberry

Commit:
shard crossref 1->2

Fixes #ISSUE_NUMBER

Regarding sharding crossref:
- Pros:
  - Does not result in a lot of additional time
  - Balances fairly well across shards
  - On linux so it's cheap
  - It is on the longer side at 2h
- Cons:
  - Not super important b/c its not the longest running job

spreadsheet regarding sharding: https://docs.google.com/spreadsheets/d/1BdtVsjRr0Is9LXMNilR02FEdPXNq7zEWl8AmR3ArsLQ/edit#gid=1153012347

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76450
Approved by: https://github.com/janeyx99, https://github.com/seemethere

renbaoshuo committed:
chore: remove git.io

All links on git.io will stop redirecting after April 29, 2022.

- https://github.blog/changelog/2022-04-25-git-io-deprecation/

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76426
Approved by: https://github.com/ngimel

Commit:
shard asan 3->4

Fixes #ISSUE_NUMBER

Regarding sharding asan more:
- Pros:
  - Does not result in a lot of additional time
  - It's on linux so its cheap
- Cons:
  - May be a bit unbalanced.  It is currently sharded into 3, with averages of 2.3h, 1.9h, 1.9h.
    - Possible solution to make it more balanced would be to break up the asan test files

spreadsheet regarding sharding: https://docs.google.com/spreadsheets/d/1BdtVsjRr0Is9LXMNilR02FEdPXNq7zEWl8AmR3ArsLQ/edit#gid=1153012347

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76404
Approved by: https://github.com/janeyx99, https://github.com/seemethere

pruthvistony committed:
Upgrade CI to ROCm 5.1

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76365
Approved by: https://github.com/seemethere, https://github.com/jeffdaily

zengk95 committed:
Revert [DataPipe] Update mux data pipe

Reverts #76384

this this is breaking tests test_demux_mux_datapipe (__main__.TestIterableDataPipeBasic. See logs: https://hud.pytorch.org/pytorch/pytorch/commit/a997046017aa9f97213f6bd50bc152a6445e8086
and was red on the PR as well: https://hud.pytorch.org/pytorch/pytorch/pull/76384
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76507
Approved by: https://github.com/kit1980

edward-io committed:
black formatting for utils/tensorboard (#76396)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/76396

Reviewed By: Reubend

Differential Revision: D35945748

Pulled By: edward-io

fbshipit-source-id: ffee22e88aaf49eb98b3e2eb6624a2dadb8ef754
(cherry picked from commit 6b5656b7c081cd69135b54f7d13d02c1c361b696)

dzdang committed:
[quant][gpu][core][bug fix] Added memset to CacheKey for quantized cudnn conv2d op (#76436)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76436

In `quantized/cudnn/Conv.cpp`, memset was added for CacheKey. Memset is needed here because there is implicit padding added for CacheKey, and this can result in uninitialized padded values that are
 used for hashing ((see how at::native::ParamsHash is defined)). without memset, we can potentially come across a situation where two CacheKey objects have the same user defined parameters, but different padded values, resulting in different hash outputs.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Reviewed By: jerryzh168

Differential Revision: D35965241

Pulled By: dzdang

fbshipit-source-id: bdeab6c3d6d6066b71b2fb313ac851fe30ae5510
(cherry picked from commit 4ac2b7a858ac62f78b49da3cf43c76d9a7371d29)

dzdang committed:
[quant][gpu][core][bug fix] Added memset to CacheKey for quantized cudnn add op (#76445)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76445

See https://github.com/pytorch/pytorch/pull/76436. It is the same idea
here.

Test Plan: python test/test_quantization.py -k test_qadd_relu_cudnn

Reviewed By: jerryzh168

Differential Revision: D35970281

Pulled By: dzdang

fbshipit-source-id: 9e983b15fa630409fa0a0906f24d94f18f1f3e7b
(cherry picked from commit 8b9be06ea7a2423e64f6ff760b1acab7423ad528)

dzdang committed:
[quant][gpu][core][bug fix] Added memset to CacheKey for quantized cudnn linear op (#76447)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76447

See https://github.com/pytorch/pytorch/pull/76436. The same idea is
applied here.

Test Plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

Reviewed By: jerryzh168

Differential Revision: D35970360

Pulled By: dzdang

fbshipit-source-id: 452f0703d3c257638d2ef1e2467756aeed0ec8d4
(cherry picked from commit ff52d26ad477bd2c0529b627ea3cc763523f80e6)

jfix71 committed:
[FX] Fix operator_schemas normalize_function to consider OpOverloads (#76469)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76469

Broken by Original commit changeset: 450e86c4e08a

Original Phabricator Diff: D35874477

Test Plan: Added unit test coverage to test_fx_experimental

Reviewed By: albanD

Differential Revision: D35978105

fbshipit-source-id: f22670b3b00a86777a26feaf4cb911595d150a17
(cherry picked from commit 91868b1e872c19d58d96a6c80a5e78dc6ffe4c7b)

anjali411 committed:
Revert "Revert "Allow specifying tags for aten operators in native_functions.yaml""

This reverts commit ea44645c9a682a4e212e64b94a86383c3388ed6b.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76456

Approved by: https://github.com/osalpekar

antoniojkim committed:
Decouple LTC from TS Backend using Lazy IR Builder

Next stage of breaking up https://github.com/pytorch/pytorch/pull/74710

IR builder class introduced to decouple the explicit usage of `TsNode` in core lazy tensors.

Requires https://github.com/pytorch/pytorch/pull/75324 to be merged in first.

**Background**
- there are ~ 5 special ops used in lazy core but defined as :public {Backend}Node.  (DeviceData, Expand, Scalar...)
- we currently require all nodes derive from {Backend}Node, so that backends can make this assumption safely
- it is hard to have shared 'IR classes' in core/ because they depend on 'Node'

**Motivation**

1. avoid copy-paste of "special" node classes for each backend
2. in general decouple and remove all dependencies that LTC has on the TS backend

**Summary of changes**
- new 'IRBuilder' interface that knows how to make 5 special ops
- move 'special' node classes to `ts_backend/`
- implement TSIRBuilder that makes the special TS Nodes
- new backend interface API to get the IRBuilder
- update core code to call the builder

CC: @wconstab @JackCaoG @henrytwo

Partially Fixes #74628

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75433
Approved by: https://github.com/wconstab

Commit:
[JIT] Add another decomposition api

Needed for functorch schema invocation
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76493
Approved by: https://github.com/davidberard98, https://github.com/Chillee

dzdang committed:
[quant][core][gpu][feature] Added support for float->quantized cuda tensor copying (#76177)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76177

Previously, support for copying a fp tensor to a quantized tensor was
limited to CPU tensors. This PR extends the support to GPU tensors.
A corresponding test was added to test_qtensor_float_assignment for cuda
tensors

Test Plan:
```
python test/test_quantization.py -k test_qtensor_float_assignment
```
Imported from OSS

Differential Revision:
D35817832
D35817832

Reviewed By: jerryzh168

Pulled By: dzdang

fbshipit-source-id: e5a4a0bb2d8a56f3f1a88806a534b5cb38275cf2
(cherry picked from commit 9173e07b51bb1b853244b205ddf3e36000f01b64)

dzdang committed:
[quant][core][gpu][improvement] Made exception throwing message clearer in Quantized Cudnn Conv2d, Linear, and Add ops (#76102)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76102

Summary plan:
Previous exception throwing messages for these 3 ops did not
differentiate amongst each other. This PR addresses that issue by
including the name of the op in the string thrown by the exception for
debugging purposes.

Test Plan:
N/A as this only changes what's written to stdout when an exception is
thrown in one of the 3 ops

Reviewed By: jerryzh168

Differential Revision: D35779082

Pulled By: dzdang

fbshipit-source-id: 4e9098995ca2b1dcd18fe294281f335740a63ec9
(cherry picked from commit 503524f3ae569f59a3b1149f68454b3b51658e3d)

wconstab committed:
Make lazy tensor creation and value strings customizable (#76472)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76472

- lets XLA backend customize codegenned functions during
  migration to LTC

Test Plan: Imported from OSS

Reviewed By: Krovatkin, bdhirsh

Differential Revision: D35980435

Pulled By: wconstab

fbshipit-source-id: 6138aef20862fccec40d715ffbb5a40a0a7d0401
(cherry picked from commit bad23f4b7ef73ffc2ef4a893364512907e9c4555)

wconstab committed:
Fix GenLazyIR.node_base_ctor_call (#76471)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76471

Make node_base_ctor_call produce the entire node_bace_ctor_call.

Previously it was only producing the beginning of the call, which was unintended.

Addresses part of https://github.com/pytorch/xla/issues/3472

Test Plan: Imported from OSS

Reviewed By: qihqi, ngimel

Differential Revision: D35980436

Pulled By: wconstab

fbshipit-source-id: a443cf593ac7c35b2b65e72b82907e88e1e71c7a
(cherry picked from commit 360ad6d82a7e8303b8a60e61b177dabf0131ea8b)

wconstab committed:
Make lazy tensor ptr class customizable (#76476)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/76476

Test Plan: Imported from OSS

Reviewed By: Krovatkin, bdhirsh

Differential Revision: D35980433

Pulled By: wconstab

fbshipit-source-id: 1d4d00a494bf8aea86278b007f7f353cd7a822f8
(cherry picked from commit a78655bef23b5fa8487ced13443ca0bfdec65e5c)

dzdang committed:
[quant][core][gpu][improvement] Converted reinterpret_cast<T *>(some_int8_tensor.data_ptr()) calls to some_int8_tensor.data_ptr<int8_t> in quantized cudnn operator files (#75980)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/75980

Support for data_ptr<T> for quantized tensor was enabled in
https://github.com/pytorch/pytorch/pull/75643. Rather than using
reinterpret_cast, we can use this overload directly. The change is
currently made in /aten/src/ATen/native/quantized/cudnn.

(Note: this ignores all push blocking failures!)

Test Plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
python test/test_quantization.py -k test_qconv2d_cudnn
python test/test_quantization.py -k test_qadd_relu_cudnn
```

Reviewed By: jerryzh168

Differential Revision: D35720654

Pulled By: dzdang

fbshipit-source-id: 5ba4b99f6cfaf1b482a0a3f5208c94e53cb05eba
(cherry picked from commit 92e2480fa0862261bff42761d5eab8ee0bb3b075)

pytorchmergebot committed:
Revert "[JIT] Add another decomposition api"

This reverts commit 32cb0375328e675d13d055a5fb1c6e2f432d0322.

Reverted https://github.com/pytorch/pytorch/pull/76493 on behalf of https://github.com/osalpekar

pytorchmergebot committed:
Revert "[JIT] Register decomp reland"

This reverts commit 81b9cb741c5d360ae51d7f214231417a1e94e7af.

Reverted https://github.com/pytorch/pytorch/pull/76397 on behalf of https://github.com/osalpekar

Commit:
[shard] Extensible Sharder and ShardingPlanner (#75844)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/75844

Add extensible Sharder and ShardingPlanner for advanced users to extend and do custom sharding, some details:

* ShardingPlanner is merely a interface that could be extended to plan advanced sharding strategies, it will produce a ShardingPlan that outlines the detailed sharding strategies for each submodule, note that it could also include custom sharders.
* a custom Sharder is another interface that could be extended to define advanced sharding strategies for specific modules (i.e. a collection of `nn.EmbeddingBag`s.) The custom sharder defined by the user can be picked up by `share_module` API. `shard_module` will call `Sharder.shard` to the module, the output of the `Sharder.shard` will replace the original submodule.

ghstack-source-id: 154978919

Test Plan:
test_sharder
test_sharding_plan

Reviewed By: pritamdamania87

Differential Revision: D35658657

fbshipit-source-id: ad9deb39a6a0ab91ceb1de3a78c4709aa50ad699
(cherry picked from commit f9796471bd7f8ad94a9d3462cac30eacacf68b31)

davidberard98 committed:
Add test_jit_cuda_fuser.py to NVFuser merge rules

Will allow NVFuser contributors to approve changes to test_jit_cuda_fuser.py

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76524

Approved by: https://github.com/janeyx99

jerryzh168 committed:
[quant] Fix tensorrt config after the backend_config_dict refactor (#76414)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76414

Previously we refactored FX Graph Mode Quantization code base to use a native backend config dict for fbgemmq/qnnpack,
because of this, we need to defien the backend config dict for tensorrt properly as well (previously it was relying on
fbgemm/qnnpack configs), this PR added some configs to enable uru10x10 again

Test Plan: buck run mode/dev-nosan -c fbcode.split-dwarf=true -c fbcode.platform=platform009 accelerators/workloads/models/uru10x10:uru_10x10_to_trt_eval -- --int8

Reviewed By: vkuzo

Differential Revision: D35939944

fbshipit-source-id: c64ade5074f5a8ee74a833bb990cd7a91c2cb152
(cherry picked from commit 02855a5ef8c196fb5b0defdfff58d6f2b94c693e)

Commit:
[PT-D][Sharding] Move Partial Tensor to the _shard folder and add logic to remove padding (#76199)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76199

Since Partial Tensor is somehow isolated to sharded tensor. We now move it to the _shard folder.

Also, we added the logic to remove paddings when the size is not divisible by the world size. Modify the unit test to reflect this changes.

Finally, we need to consider the placement order for the reshading spec for partial tensor, related logic is added in this change. Futhermore, for sharded linear, we will need to order the placement by rank to get the expected local result.
ghstack-source-id: 154853290

Test Plan: CI

Reviewed By: pritamdamania87, wanchaol

Differential Revision: D35827894

fbshipit-source-id: 58dab77969b8b6557f42afa7e8f5a8a053dd5793
(cherry picked from commit abeb28f16582dcf707c2e165f39df6caf692384d)

kevinstephano committed:
Adding `broadcast_in_dim` and non-contiguous Tensor support NVFuser Python Frontend

Adding new features.

1. `broadcast_in_dim` support and example.
2. Adding non-contiguous `TensorView` support and example.

`broadcast_in_dim` example:
```
with FusionDefinition(fusion) as fd :
    t0 = fd.define_tensor(1)
    t1 = fd.define_tensor(3)

    fd.add_input(t0)
    fd.add_input(t1)

    t0_b = fd.Ops.broadcast_in_dim(t0, [2, 3, 4], [1])
    t2 = fd.Ops.add(t0_b, t1)

    fd.add_output(t2)
```

Non-contiguous tensor support example:

```
with FusionDefinition(fusion) as fd :
    t0 = fd.define_tensor(3, [False, False, False])
    t1 = fd.define_tensor(3, [True, True, True])

    fd.add_input(t0)
    fd.add_input(t1)
    print("Input1 Contiguity:", t0)
    print("Input2 Contiguity:", t1)

    t2 = fd.Ops.add(t0, t1)

    print("Output Contiguity:", t2, "\n")
    fd.add_output(t2)
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76484
Approved by: https://github.com/mruberry

dzdang committed:
[quant][core][gpu][feature] Implemented quantized cuda adaptive average pool2d op (#76081)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76081

The current implementation of quantized cuda adaptive average pooling uses the following:
dequant -> fp32 adaptive average pooling -> quant. This is the same numerically as quantized adaptive average pooling. This is not the ideal implementation, as we desire to operate on the quantized values directly. However, we are currently blocked on this as we are waiting for cudnn's 8.5.0 release, which is anticipated to support adaptive average pooling. When that support is made available, we will use it directly.

Test Plan:
```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

Differential Revision:
D35768751
D35768751

Reviewed By: jerryzh168

Pulled By: dzdang

fbshipit-source-id: ad06fd06d6941b92105bcabf0fd54b9e27a029d5
(cherry picked from commit 4e1805dd62a9d5e94c61340ac46bcd7aa4e49dd9)

rraminen committed:
Hipify fixes for a successful DeepSpeed build

These commits are required to build DeepSpeed on ROCm without the hipify errors.

https://github.com/ROCmSoftwarePlatform/pytorch/commit/a41829d9edd576775b68ca330799158960a8b766
https://github.com/ROCmSoftwarePlatform/pytorch/commit/663c7184629a0a721df540b74bcb78b1a66d0310

cc: @jeffdaily

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76141
Approved by: https://github.com/jeffdaily, https://github.com/pruthvistony, https://github.com/albanD

dzdang committed:
[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator

Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76184

Approved by: https://github.com/jerryzh168

janeyx99 committed:
Disable XLA test workflow

Our trunk is hosed rn https://hud.pytorch.org/minihud?name_filter=pull%20/%20pytorch-xla-linux-bionic-py3.7-clang8%20/%20test%20(xla,%201,%201,%20linux.2xlarge)

This mitigates by disabling the workflow
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76537
Approved by: https://github.com/atalman

janeyx99 committed:
Shard rocm distributed tests

As ROCm enabled more distributed tests on trunk, the test time has increased dramatically and is now past the 5hr timeout threshold.

This is a mitigation! Note that distributed tests take 2.5 hrs ish on a linux machine, but they take over 5 hours now on ROCm.
One can observe by going through the pages in https://hud.pytorch.org/hud/pytorch/pytorch/master/3?name_filter=rocm and seeing that after fdsp tests were enabled + some other changes, ROCm dist test time increased by 1hr+.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76536
Approved by: https://github.com/zengk95

rdspring1 committed:
Nvfuser - Type Promotion Fix

Fix Type Promotion failures in [Issue 76046](https://github.com/pytorch/pytorch/issues/76046)

1. Updated nvfuser type promotion rule for codegen kernel;
2. Updated casting for output of nvfuser kernel to respect profiling/TorchScript scalar type;
3. Updated type_inference.cpp to only update device/scalar_type when profiling information is missing.

Additional Type Promotion Fixes:
-  test_nvfuser_correctness_softmax_with_dtype_cuda_float32
-  test_nvfuser_correctness_softmax_with_dtype_cuda_bfloat16
-  test_nvfuser_correctness_softmax_with_dtype_cuda_float16
-  test_nvfuser_correctness_softmax_with_dtype_cuda_float32
-  test_nvfuser_correctness_log_softmax_dtype_cuda_bfloat16
-  test_nvfuser_correctness_log_softmax_dtype_cuda_bool
-  test_nvfuser_correctness_log_softmax_dtype_cuda_float16
-  test_nvfuser_correctness_log_softmax_dtype_cuda_float32
-  test_nvfuser_correctness_sum_cuda_int32
-  test_nvfuser_correctness_sum_to_size_cuda_int32
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76343
Approved by: https://github.com/jjsjann123, https://github.com/mruberry

jbschlosser committed:
Fix string repr for nested tensors & run nested tensor tests

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76541

Approved by: https://github.com/ishaan-mehta, https://github.com/cpuhrsch

albanD committed:
Readme update to remove old python version

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76328
Approved by: https://github.com/ptrblck, https://github.com/seemethere

justinchuby committed:
[ONNX] Documentation and typing annotations in registry

Updating the docstrings and type annotations as I walk through the code.

- Turned some comments into docstrings.
- Added type annotations for some functions in utils and the registry
- Removed direct function imports; importing functions makes name space collision easier to happen and refactoring/code analysis harder: https://google.github.io/styleguide/pyguide.html#22-imports
- Formatted touched files with black
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76255
Approved by: https://github.com/BowenBao

IvanYashchuk committed:
Add torch.linalg.ldl_factor_ex and torch.linalg.ldl_solve

This PR adds a function for computing the LDL decomposition and a function that can solve systems of linear equations using this decomposition. The result of `torch.linalg.ldl_factor_ex` is in a compact form and it's required to use it only through `torch.linalg.ldl_solve`. In the future, we could provide `ldl_unpack` function that transforms the compact representation into explicit matrices.

Fixes https://github.com/pytorch/pytorch/issues/54847.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano
Pull Request resolved: https://github.com/pytorch/pytorch/pull/69828
Approved by: https://github.com/Lezcano, https://github.com/mruberry, https://github.com/albanD

Commit:
[FSDP] Relax exec order valid. to only fwd

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76556

Approved by: https://github.com/zhaojuanmao, https://github.com/rohan-varma

janeyx99 commented on .github/scripts/trymerge.py:
```diff
@@ -652,6 +654,25 @@ def merge_into(self, repo: GitRepo, *, force: bool = False, dry_run: bool = Fals
+            try:
+                self.merge_into(repo, dry_run=dry_run)
+            except Exception as e:
+                print(f"Merge failed due to {e}. Trying again in 60 seconds.")
```
Shouldn't we then check if it failed for a legitimate reason? I think this is a good place to parse the reason and if it's something other than waiting for status checks, it should fail and not keep while looping.

zengk95 replied:
oh good point. I guess we only need to try again if it's a check exception and stop if there's any other types.

pearu committed:
Unsafe Sparse Compressed tensor factory function

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75961

Approved by: https://github.com/cpuhrsch

seemethere commented on .github/scripts/trymerge.py:
```diff
@@ -652,6 +654,25 @@ def merge_into(self, repo: GitRepo, *, force: bool = False, dry_run: bool = Fals
+    def merge_on_green(self, repo: GitRepo, org: str, project: str, pr_num: int, *, dry_run: bool = False) -> None:
+        start_time = time.time()
+
+        while True:
```
Wait does this mean that the merge job will just continually run until the conditions are met?

Shouldn't we just base the merge on green logic on webhook events to make it simpler?

seemethere replied:
Also for some of the attempted merges couldn't we potentially get into land races with other PRs?

I'm also skeptical if Github would accept a push that's out of date with the current master branch as well

zengk95 replied:
> Wait does this mean that the merge job will just continually run until the conditions are met?
Yes that's the plan is to just continually call try-merge until we get it working

> Shouldn't we just base the merge on green logic on webhook events to make it simpler?
that was  the original plan and I implemented that in here but got some feedback to try to just put it in here so that it's all in one place to keep it more simple. https://github.com/pytorch/test-infra/pull/302

zhaojuanmao committed:
Provide an auto wrap policy for common transformer models

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76455

Provide an auto wrap policy for common transformer models

Differential Revision: [D35972488](https://our.internmc.facebook.com/intern/diff/D35972488/)

Approved by: https://github.com/rohan-varma

davidberard98 committed:
Retry - [NVFuser] always use fallback if fusion fails

Retry of #75983. The change is to handle cases where attr::cache_id is
not set. This can happen if compilation fails.

Original message:

1) remember when fusions fail; and on subsequent runs, always take the fallback.
2) during the first fallback, cache the Code object.

On autogen-69 from the nvfuser microbenchmarks (https://github.com/pytorch/benchmark/pull/801) this improved performanance as follows:
* Original (always attempt fusion): 25ms
* Always take fallback after first failure: 0.79ms
* Always take fallback + cache Code object: 0.62ms
* Eager: 0.58ms

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76505
Approved by: https://github.com/eellison

Commit:
Reland reland

Reland of https://github.com/pytorch/pytorch/pull/76397 and https://github.com/pytorch/pytorch/pull/76493

This time I'll get it right 😢
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76539
Approved by: https://github.com/davidberard98, https://github.com/osalpekar

janeyx99 committed:
[GH1] Fix pending/not run conditional and add a way to test new merge_rules

Addressing confusing developer experience here https://github.com/pytorch/pytorch/pull/69828#issuecomment-1112451260

Also adds a way to test merge rules.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76559
Approved by: https://github.com/seemethere, https://github.com/albanD

Commit:
Export aten::to("cpu") and aten::to(device="cpu")

Fixes https://github.com/facebookresearch/detectron2/pull/4120 after https://github.com/facebookresearch/detectron2/pull/4132 was merged

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76498
Approved by: https://github.com/garymm

jjsjann123 committed:
Add a matching lerp implementation to eager mode. (#1612)

Fixes part of #76046

Add a matching lerp to eager mode.

Co-authored-by: jjsjann123 <alex.jann2012@gmail.com>
Co-authored-by: jjsjann123 <jiej@nvidia.com>
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76459
Approved by: https://github.com/ngimel

chunyuan-w committed:
[Re-landing 68111] Add JIT graph fuser for oneDNN Graph API (Preview4.1)

Re-landing https://github.com/pytorch/pytorch/pull/68111

## Description
Preview4 PR of this [RFC](https://github.com/pytorch/pytorch/issues/49444).

On the basis of https://github.com/pytorch/pytorch/pull/50256, the below improvements are included:

- The [preview4 release branch](https://github.com/oneapi-src/oneDNN/releases/tag/graph-v0.4.1) of the oneDNN Graph API is used
- The fuser now works with the profiling graph executor. We have inserted type check nodes to guard the profiled tensor properties.

### User API:
The optimization pass is disabled by default. Users could enable it by:
```
torch.jit.enable_onednn_fusion(True)
```

### Performance:
[pytorch/benchmark](https://github.com/pytorch/benchmark) tool is used to compare the performance:
- SkyLake 8180 (1 socket of 28 cores):

  ![image](https://user-images.githubusercontent.com/65992142/151162305-05e44425-a24e-4d5e-94e1-743b40b87a8c.png)

- SkyLake 8180 (single thread):

  ![image](https://user-images.githubusercontent.com/65992142/151162528-69f90b79-d08d-46b8-8775-d80a6ccbce8a.png)
 \* By mapping hardswish to oneDNN Graph, it’s 8% faster than PyTorch JIT (NNC + OFI)
  \** We expect performance gain after mapping transpose, contiguous & view to oneDNN graph ops

### Directory structure of the integration code
Fuser-related code are placed under:
```
torch/csrc/jit/codegen/onednn/
```

Optimization pass registration is done in:
```
torch/csrc/jit/passes/onednn_graph_fuser.h
```

CMake for the integration code is:
```
caffe2/CMakeLists.txt
```

## Limitations

- In this PR, we have only supported the optimization on Linux platform. The support on Windows and MacOS will be enabled as the next step.
- We have only optimized the inference use case.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/74596
Approved by: https://github.com/malfet

JackCaoG committed:
Enable xla test

Reenable the test since offending pr is reverted.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76565
Approved by: https://github.com/janeyx99

Commit:
[primTorch] Prototype nvFuser integration and test_prims.py

This adds prototype nvFuser integration for the following prims:

- broadcast_in_dim
- convert_element_type
- add
- div
- ge
- gt
- le
- lt
- mul

Adding it for additional prims supported by nvFuser's prototype Python frontend should be easy.

This also adds a new sugar to run operations using the ATen or nvFuser trace executors. For example:

```
def foo(a, b):
  return torch.add(a, b)

traced_foo = make_traced(foo)

a = torch.randn((1, 2, 3, 4, 5), device='cuda')
b = torch.randn((1, 2, 3, 4, 5), device='cuda')
result = traced_foo(a, b, executor='nvfuser')
```

Currently only operations with tensor inputs and one tensor output are supported, and the operation must be composed exclusively of reference or prim operations.

Finally, this adds a new test, test_prims.py, that just tests the broadcast_in_dim prim for now. In the future we'll likely have OpInfos for each prim, but we'll need a reference implementation of broadcast_in_dim to make that interesting.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76560
Approved by: https://github.com/ngimel

janeyx99 committed:
Allow sharding for distributed tests

Addresses my mistake introduced in https://github.com/pytorch/pytorch/pull/76536#issuecomment-1112657429

Also allows for sharding 1 in run_test.py
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76570
Approved by: https://github.com/jeffdaily, https://github.com/seemethere

lezcano committed:
Add linalg.vander

This PR adds `linalg.vander`, the linalg version of `torch.vander`.

We add autograd support and support for batched inputs.

We also take this chance to improve the docs (TODO: Check that they
render correctly!) and add an OpInfo.

**Discussion**: The current default for the `increasing` kwargs is extremely
odd as it is the opposite of the classical definition (see
[wiki](https://en.wikipedia.org/wiki/Vandermonde_matrix)). This is
reflected in the docs, where I explicit both the odd defaults that we
use and the classical definition. See also [this stackoverflow
post](https://stackoverflow.com/a/71758047/5280578), which shows how
people are confused by this defaults.

My take on this would be to correct the default to be `increasing=True`
and document the divergence with NumPy (as we do for other `linalg`
functions) as:

- It is what people expect
- It gives the correct determinant called "the Vandermonde determinant" rather than (-1)^{n-1} times the Vandermonde det (ugh).
- [Minor] It is more efficient (no `flip` needed)
- Since it's under `linalg.vander`, it's strictly not a drop-in replacement for `np.vander`.

We will deprecate `torch.vander` in a PR after this one in this stack
(once we settle on what's the correct default).

Thoughts? mruberry

cc kgryte rgommers as they might have some context for the defaults of
NumPy.

Fixes https://github.com/pytorch/pytorch/issues/60197

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76303

Approved by: https://github.com/albanD

pytorchmergebot committed:
Revert "[Re-landing 68111] Add JIT graph fuser for oneDNN Graph API (Preview4.1)"

This reverts commit 8b11d810583ab1aac16b211efcc131c85d17c502.

Reverted https://github.com/pytorch/pytorch/pull/74596 on behalf of https://github.com/janeyx99

fps7806 committed:
[CUDA Graphs] Fix OOM inside graph capture_begin

release_cached_blocks calls this:
```
void synchronize_and_free_events() {
    TORCH_INTERNAL_ASSERT(captures_underway == 0);
```
Which means we can't call that function when we are capturing a cuda graph:
```
import torch

with torch.cuda.graph(torch.cuda.CUDAGraph()):
    torch.zeros(2 ** 40, device="cuda")
```

results in:
```
RuntimeError: captures_underway == 0INTERNAL ASSERT FAILED at "/tmp/torch/c10/cuda/CUDACachingAllocator.cpp":1224, please report a bug to PyTorch.
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76247
Approved by: https://github.com/ngimel

Chillee committed:
Add better error message for FX when using concrete_args

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76600
Approved by: https://github.com/ZolotukhinM

francescocastelli committed:
Make Opinfo a dataclass to get a more useful repr

Fixes #74320. If we make OpInfo a dataclass we should change the method `default_test_dtypes` name because there is also an argument with the same name (or change the argument name). Right now I've changed the method name because it seems to be used only twice.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75663
Approved by: https://github.com/mruberry

jjsjann123 committed:
Nvfuser faster fallback

Follow up to #76505

Addressing https://github.com/pytorch/pytorch/pull/76505#discussion_r861260818 to further improve fallback perf during compilation failure. This allows us to reuse fallback instead re-constructing new code every time.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76604
Approved by: https://github.com/davidberard98

kshitij12345 committed:
[fix] mul chalf: cpu scalar and cuda tensor case

https://github.com/pytorch/pytorch/pull/76158 added `chalf` support for `mul` on CUDA incorrectly.

Following sample
```python
torch.mul(torch.zeros(3, device='cuda'), 2.5) # CUDA Tensor and CPU Scalar
```

fails with
```
RuntimeError: iter.device(arg).is_cuda() INTERNAL ASSERT FAILED at "../aten/src/ATen/native/cuda/JitLoops.cuh":83, please report a bug to PyTorch. argument 2: expected a CUDA device but found cpu
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76364
Approved by: https://github.com/mruberry

ezyang committed:
Move functorch decompositions to PyTorch

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76311

Approved by: https://github.com/Chillee

Chillee committed:
Port functorch decomps over and fix some tests

Still some stuff to fix up, will finish later.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76621
Approved by: https://github.com/ezyang

kshitij12345 committed:
[complex32] support printing the tensor

Reference: #74537
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76614
Approved by: https://github.com/anjali411

rohan-varma committed:
Change mixed precision API

Change the API such that if a precision type is not explicitly speciifed in `MixedPrecision` config, we don't cast it at all.

Previously, we would default to fp16 reduced precision, but this doesn't account for the case where user might want to, for example, use only reduced gradient comm precision. Trying to do this via:
```
MixedPrecision(reduce_dtype=torch.float16, param_dtype=torch.float32, buffer_dtype=torch.float32)
```

does not work for all use cases because the code will still attempt to cast params to fp32, but user's model may be assuming double type/fp64 somewhere.

Now, specifying
```
MixedPrecision(reduce_dtype=torch.float16)
```

only affects gradient comm precision, and does not touch casting of params / buffers.

Note that if user specifies reduced precision for only parameters, gradients will be of this reduced type in _post_backward_hook and are therefore communicated in this precision. Therefore, the priority of precision in which grads are communicated is:

reduce_dtype, if specified -> param_dtype, if specified -> full precision param type.

We take additional care to make sure grads are cast back to the full precision param type for optimizer step: either if parameter was in reduced precision or if parameter was not in reduced precision but reduced gradient precision for comm was configured.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76423
Approved by: https://github.com/zhaojuanmao

rdspring1 committed:
[Nvfuser] Add cast support between double and half types

Fixes `RuntimeError: Illegal Cast value from  DataType: __half to DataType: double`

Example:
```
with FusionDefinition(fusion) as fd :
    t0 = fd.define_tensor(2, DataType.Half)
    t1 = fd.define_tensor(2, DataType.Double)

    fd.add_input(t0)
    fd.add_input(t1)

    t2 = fd.Ops.add(t0, t1)
    t5 = fd.Ops.relu(t2)

    fd.add_output(t5)
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76605
Approved by: https://github.com/csarofeen

Commit:
Adds python ref consistency test, elementwise unary reference inputs, and formats test files

Per title.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76626
Approved by: https://github.com/ngimel

Commit:
Change 'python mode' to 'torch dispatch mode'

Changes Python Mode name to Torch Dispatch Mode because there is now a Torch Function Mode, so Torch Dispatch Mode and Torch Function Mode are consistent with each other
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76562
Approved by: https://github.com/zou3519

IvanYashchuk committed:
Fix failing test when SciPy is not available for test_ldl_factor

This PR moves the SciPy function import under the `if TEST_SCIPY` block.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76657
Approved by: https://github.com/nikitaved

bhashemian committed:
Fix lists in the docstring

Signed-off-by: Behrooz <3968947+drbeh@users.noreply.github.com>

Fixes #76593
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76594
Approved by: https://github.com/ejguan

vitrioil committed:
Removed direct doc formatting

Fixes #76034

This does not make python remove all `__doc__` because in some places `__doc__` is assigned to a string.

Example:
https://github.com/pytorch/pytorch/blob/04b3313379712098183dfe5bea002a5e43b5af48/torch/nn/modules/conv.py#L174-L233

Since there are quite a few of these, I will add all of them together in this PR later. (Basically still a lot of docstring will persist even with `-OO` enabled.)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76619
Approved by: https://github.com/albanD

Commit:
adds elementwise opinfos and unary references, extends to out testing

This PR makes the following changes:

Prims:
- igamma and igammac are now correctly listed as elementwise binary operations, not elementwise unary operations
- elementwise prims now must specify their type promotion kind (this is currently unused)

Refs:
- complexhalf is now handled by opmath-style type promotion
- adds references for: abs, acos, acosh, asin, atan, ceil, cos, cosh, digamma, erf, erfinv, erfc, exp, expm1, isfinite, isnan, lgamma, log, log1p, neg, reciprocal, sign, sin, sinh, sqrt, square, tan, igamma, igammac
- adds "complex to float" and "bool to long" type promotion kinds
- updates out behavior to warn when resizing a non-empty tensor, consistent with current ops
- updates the elementwise unary reference template with type promotion

Tests:
- fixes torch.pow's OpInfo to correctly specify it only supports one scalar input, not two
- fixes elementwise binary reference inputs to not attempt generating certain tensors in complex half (for now, cc @kshitij12345)
- adds OpInfos for the following Python references: abs, acos, acosh, asin, atan, ceil, cos, cosh, digamma, erf, erfinv, erfc, exp, expm1, isfinite, isnan, lgamma, log, log1p, neg, reciprocal, round, sign, sin, sinh, sqrt, square, tan, atan2, bitwise_and, bitwise_left_shift, bitwise_or, bitwise_xor, eq, float_power, ge, gt, igamma, igammac, le, lt, maximum, minimum, mul, ne, nextafter, pow, sub, true_divide

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76647
Approved by: https://github.com/ngimel

AdityaKane2001 committed:
Nit in `TripletMarginLoss`

`:math` -> `:math:`

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76629
Approved by: https://github.com/albanD

pytorchmergebot committed:
Revert "Change 'python mode' to 'torch dispatch mode'"

This reverts commit 7203a73986943c50b8d2f0fdc9d09821705a1602.

Reverted https://github.com/pytorch/pytorch/pull/76562 on behalf of https://github.com/janeyx99

peterbell10 committed:
Cleanup ATen-core forward declarations

I noticed that when `SymInt` was introduced, `jit_type_base.h` was
added as an include to the `Operator.h` template which is supposed to
be kept extremely clean and only use forward declarations. Also,
that forward declarations for `OptionalArrayRef` were missing.

So, I've refactored the forward declarations into
`ATen/core/ATen_fwd.h` and cleaned up some of the `c10`
headers that were masking these missing declarations. I've also
re-generated the pre-compiled header so `SymInt` is included.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76576
Approved by: https://github.com/albanD

zengk95 committed:
actually add it

zengk95 committed:
actually add it

janeyx99 commented on .github/scripts/trymerge.py:
```diff
@@ -735,22 +739,29 @@ def find_matching_merge_rule(pr: GitHubPR,
-                    pass_checks = False
-            if not pass_checks:
-                continue
+                if checkname not in checks or checks[checkname]:
```
```suggestion
                if checkname not in checks or checks[checkname] == null:
```
missing conditional?

janeyx99 commented on .github/scripts/trymerge.py:
```diff
@@ -735,22 +739,29 @@ def find_matching_merge_rule(pr: GitHubPR,
+            if reject_reason_score < 20000:
+                reject_reason_score = 20000
+                reject_reason = f"Refusing to merge as mandatory check(s) {','.join(pending_checks)}"
+                reject_reason += f"are pending for rule {rule_name}"
```
Or not run at all, right?
```suggestion
                reject_reason += f"are not yet run for rule {rule_name}"
```

janeyx99 commented on .github/scripts/trymerge.py:
```diff
@@ -735,22 +739,29 @@ def find_matching_merge_rule(pr: GitHubPR,
             raise RuntimeError("This PR has internal changes and must be landed via Phabricator")
         return rule
+    if reject_reason_score == 20000:
+        raise MandatoryChecksError(reject_reason)
```
Could we call this something different because wouldn't error == 30000 also technically be a mandatorycheckserror?

janeyx99 reviewed (commented):
Test plan?

And agree we should look out for the potential errors that Eli mentions, such as needing to repull master etc.

suo commented on .github/scripts/trymerge.py:
```diff
@@ -834,6 +865,17 @@ def main() -> None:
+        except Exception as e:
+            handle_exception(e)
+
+    try:
```
this needs to be in an `else` branch otherwise we will try to merge twice

suo commented on .github/scripts/trymerge.py:
```diff
@@ -797,6 +808,28 @@ def post_comment(msg: str) -> None:
+        try:
+            pr.merge_into(repo, dry_run=dry_run)
+        except MandatoryChecksError as e:
+            print('Merged failed due to {e}. Retrying in 60 seconds.')
```
Might be useful to attach the pending check name to the exception and print it here for debugging purposes.

zengk95 committed:
resolved comments

zengk95 committed:
merge

zengk95 commented:
whoops. I messed up the git history and accidentally tagged a bunch of people
